applications/HDF5Application/custom_io/hdf5_file.h

```diff
@@ -85,216 +99,409 @@ class KRATOS_API(HDF5_APPLICATION) File
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    virtual void ReadDataSet(const std::string& rPath,
-                             Matrix<int>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    virtual void ReadDataSet(const std::string& rPath,
-                             Matrix<double>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    // Independently read data set from the HDF5 file.
     /**
-     *  Performs independent read in MPI. Throws if out of range.
+     * @brief Independently read a data set from the HDF5 file.
+     *
+     * Performs independent read in MPI. Throws if out of range.
+     *
+     * @tparam TDataType        Datatype of the read data.
+     * @param rPath             Path of the dataset.
+     * @param rData             Data to be written to.
+     * @param StartIndex        Starting offset of data for this rank.
+     * @param BlockSize         Number of data points for this rank.
      */
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<int>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<double>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<array_1d<double, 3>>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Matrix<int>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Matrix<double>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
+    template<class TDataType>
+    void ReadDataSetIndependent(
+        const std::string& rPath,
+        TDataType& rData,
+        const unsigned StartIndex,
+        const unsigned BlockSize);
 
     unsigned GetOpenObjectsCount() const;
     ///@}
 
-protected:
-    ///@name Protected Operations
-    ///@{
-    hid_t GetFileId() const;
-    ///@}
-
 private:
     ///@name Member Variables
     ///@{
+
+    DataCommunicator const * mpDataCommunicator;
+
     std::string m_file_name;
+
     hid_t m_file_id = -1; // Default invalid file id.
+
     int m_echo_level = 0;
```

Comment:
Now's the chance to fix the naming.


```suggestion
    std::string mFileName;

    hid_t mFileId = -1; // Default invalid file id.

    int mEchoLevel = 0;
```

Reply:
+ a minor nitpick about the order of member variables:

In general, there are 2 aspects to keep in mind when deciding on the order of member variables:
- memory alignment and padding (not too important in this case because the order of variables won't have much effect; just don't put the `std::string` in the middle)
- memory access patterns; i.e.: which members get accessed most often? You can improve your chances of hitting the cache if you put your most often used member vars in the front (`m_file_id`).

Reply:
done :)